tests/robustness/coverage/key_pattern_test.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2025 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coverage_test
+
+import (
+	"strings"
+	"testing"
+)
+
+func pattern(key string) (string, bool) {
+	if key == "/registry/health" || key == "compact_rev_key" {
+		return key, true
+	}
+	key, valid := strings.CutPrefix(key, "/registry")
+	if !valid {
+		return key, false
+	}
+	slash := ""
+	if strings.HasSuffix(key, "/") {
+		slash = "/"
+	}
+	parts := strings.Split(strings.Trim(key, "/"), "/")
+	if len(parts[0]) == 0 {
+		return key, false
+	}
+
+	apiResource, shift := resourceOrCRD(parts[0])
+	parts = parts[shift:]
+
+	switch len(parts) {
+	case 0:
+		// Listing all resources.
+		return "/registry/" + apiResource + slash, true
+	case 1:
+		if slash == "" {
+			// Get on a non-namespaced resource.
+			return "/registry/" + apiResource + "/{name}", true
+		}
+		// Listing all resources in a namespace.
+		return "/registry/" + apiResource + "/{namespace}" + slash, true
+	case 2:
+		if slash == "" {
+			// Get on a namespaced resource.
+			return "/registry/" + apiResource + "/{namespace}/{name}", true
+		}
+	}
+
+	return key, false
+}
+
+func resourceOrCRD(s string) (string, int) {
+	if strings.Contains(s, ".") {
+		// Group name from the Custom Resource Definition.
+		// Names without dots are technically valid DNS name, but against best practices:
```

Comment:
So are you saying there is no guarantee that there is a "." in the CRD and you might incorrectly classify a pattern?

Reply:
This is correct, it is possible that some non-namespaced CRD resources will be classified as `/registry/{resource}/{namespace}/{name}` instead of `/registry/{api-group}/{resource}/{name}`, but the other cases should show up in the list as a new (unexpected) pattern

Reply:
Makes sense, thanks for confirming.